File: swift/133_CloneGraph.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var neighbors: [Node?]
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.neighbors = []
 *     }
 * }
 */

class Solution {
    var clonedNodesMap = [Node? : Node?]()
    func cloneGraph(_ node: Node?) -> Node? {
        if node == nil { return nil }
        let clone: Node? = Node(node!.val)
        clonedNodesMap[node] = clone
        return dfs(node, clone)
    }
    
    func dfs(_ node: Node?, _ clone: Node?) -> Node? {
        var clone = clone
        for neighbor in node!.neighbors {
            if let clonedNeighbor = clonedNodesMap[neighbor] {
                clone!.neighbors.append(clonedNeighbor)
            } else {
                let clonedNeighbor: Node? = Node(neighbor!.val)
                clonedNodesMap[neighbor] = clonedNeighbor
                clone!.neighbors.append(clonedNeighbor)
                dfs(neighbor, clonedNeighbor)
            }
        }
        return clone
    }
}

File: java/133_CloneGraph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    HashMap<Node, Node> clonedNodesMap;
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        
        Node clone = new Node(node.val);
        clonedNodesMap = new HashMap<Node, Node>();
        clonedNodesMap.put(node, clone);
        
        return dfs(node, clone);
    }
    
    public Node dfs(Node node, Node clone) {
        for (Node nei : node.neighbors) {
            if (clonedNodesMap.get(nei) == null) {
                Node clonedNeighbor = new Node(nei.val);
                clonedNodesMap.put(nei, clonedNeighbor);
                clone.neighbors.add(clonedNeighbor);
                dfs(nei, clonedNeighbor);
            } else {
                clone.neighbors.add(clonedNodesMap.get(nei));
            }
        }
        return clone;
    }
}
